libs/partners/cohere/langchain_cohere/react_multi_hop/agent.py

```diff
@@ -0,0 +1,317 @@
+                regex = r"^(Action\s*\d*\s*:\s*\d*\s*```json\n.*?```)"
+                action_match = re.search(regex, generation, re.DOTALL)
+                if not action_match:
+                    raise ValueError(
+                        f"Failed to parse multihop completion for input: {generation}"
+                    )
+                actions = action_match.group(1).strip()
+        except Exception as e:
+            logging.error(
+                f"Failed to parse multihop completion for input: {generation}"
+            )
+            logging.error(f"Error: {e}")
+
+        parsed_actions = self.parse_jsonified_tool_use_generation(actions, "Action:")
+        return generation, plan, parsed_actions
+
+    def parse(self, text: str) -> Union[List[AgentAction], AgentFinish]:
+        # Parse the structured output of the final answer.
+        if "Answer: " in text:
+            prefix_map = {
+                "answer": "Answer:",
+                "grounded_answer": "Grounded answer:",
+                "relevant_docs": "Relevant Documents:",
+                "cited_docs": "Cited Documents:",
+            }
+            parsed_answer = self.parse_answer_with_prefixes(text, prefix_map)
+            return AgentFinish({"output": parsed_answer["answer"]}, text)
+        elif any([x in text for x in ["Plan: ", "Reflection: ", "Action: "]]):
+            completion, plan, actions = self.parse_actions(text)
+            return [
+                AgentAction(
+                    action["tool_name"],
+                    action["parameters"],
+                    f"\n{action}\n" if i > 0 else f"\n{plan}\n{action}\n",
+                )
+                for i, action in enumerate(actions)
+            ]
+        else:
+            raise ValueError(
+                "\nCould not parse generation as it did not contain Plan, Reflection,"
+                + f"Action, or Answer. Input: {text}\n\n"
+            )
+
+
+def get_type(type_: str, is_optional: bool) -> str:
+    python_type = JSON_TO_PYTHON_TYPES.get(type_, type_)
+    if is_optional:
+        return f"Optional[{python_type}]"
+    else:
+        return python_type
+
+
+def get_tool_signature(tool: BaseTool) -> str:
+    """Get the tool signature."""
+    args = []
+    for parameter_name, parameter_definition in tool.args.items():
+        type_ = get_type(
+            parameter_definition.get("type"), "default" in parameter_definition
+        )
+        args.append(f"{parameter_name}: {type_}")
+    signature = ", ".join(args)
+    return f"def {tool.name}({signature}) -> List[Dict]:"
+
+
+def get_tool_args(tool: BaseTool) -> str:
+    """Get the tool args."""
+    if not tool.args:
+        return ""
+    rendered_args = []
+    for parameter_name, parameter_definition in tool.args.items():
+        type_ = get_type(
+            parameter_definition.get("type"), "default" in parameter_definition
+        )
+        description = parameter_definition.get("description")
```

Comment:
```suggestion
        description = parameter_definition.get("description") or ""
```
to avoid rendering `None` into param description